### PR TITLE
Make Deflate/GZip/ZlibStream empty writes have the same impact as no writes

### PR DESCRIPTION
**Pull request by stephentoub**
If you create such a compression stream, don't write to it, and close it, it won't write out any bytes.  But today if you perform an empty write, it'll result in the stream thinking data was written, and it'll end up writing out an empty chunk, which will result in several bytes of output.  Logically it's strange that there's such a difference between an empty write and no write; this seeks to make them the same.

This changed back in 2015, in https://github.com/dotnet/corefx/pull/4917, I _think_ unintentionally.

@ericstj, this removes a test you added in https://github.com/dotnet/corefx/pull/41190 (the test relies on an empty write resulting in a non-empty compressed payload).  Do you have concerns with this change? Is there reason to believe folks depend on this?  In general with compression we make no guarantees version to version about generated bytes other than that they'll roundtrip, but this change obviously breaks that test you added, which will now loop forever because it assumes the number of bytes written will not be 0.

### CODE REVIEW

**Comment by ericstj** [comment]
The main reason I added that test was so that I could reliably produce cases where the `Inflate` would return 0 bytes when not finished - to test the fix made to copy-to that mis-interpreting that.  It would be nice if we still had a way to produce those empty chunks since it's a valuable test case.  Could we do that by just having the test write the raw byte sequence for an empty chunk instead?

I agree that someone might _notice_ that zero-byte writes produce nothing when they previously produced these empty chunks - but I agree the new behavior is more desirable.  Zips with empty files will change with this - for the better.  Agreed that https://github.com/dotnet/corefx/pull/4917 changed this unintentionally.

File: src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs [hunk 1]
```diff
@@ -55,64 +55,6 @@ private class DerivedMemoryStream : MemoryStream
...
-                {
-                    // write one byte in order to allow us to prime the inflater buffer
-                    gz.WriteByte(3);
-                }
-
-                while (compressedStream.Length < minCompressedSize)
-                {
-                    using (var gz = new GZipStream(compressedStream, CompressionLevel.NoCompression, leaveOpen: true))
-                    {
-                        gz.Write(Array.Empty<byte>());
```
Comment (ericstj) [inline src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs:88]: Could we preserve this test by writing raw bytes to `compressedStream`?

Comment (stephentoub) [reply src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs:88]: Done

**Comment by carlossanlop** [comment]
>  In general with compression we make no guarantees version to version about generated bytes other than that they'll roundtrip

> I agree that someone might notice that zero-byte writes produce nothing when they previously produced these empty chunks - but I agree the new behavior is more desirable

Should we mark this as a breaking change?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs [hunk 2]
```diff
@@ -507,8 +507,13 @@ internal void WriteCore(ReadOnlySpan<byte> buffer)
             EnsureCompressionMode();
             EnsureNotDisposed();
 
-            Debug.Assert(_deflater != null);
+            if (buffer.IsEmpty)
+            {
+                return;
+            }
+
             // Write compressed the bytes we already passed to the deflater:
```
Comment (carlossanlop) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:515]: I know it's not from your change, but shouldn't this message say this?:

```suggestion
            // Write the compressed bytes we already passed to the deflater:
```

File: src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs (hunk 1 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/tests/CompressionStreamUnitTests.Gzip.cs:88]: Nice case.

**Review by carlossanlop — APPROVED** [review]

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs [hunk 3]
```diff
@@ -793,8 +798,9 @@ internal ValueTask WriteAsyncMemory(ReadOnlyMemory<byte> buffer, CancellationTok
             EnsureNoActiveAsyncOperation();
             EnsureNotDisposed();
 
-            return cancellationToken.IsCancellationRequested ?
-                ValueTask.FromCanceled(cancellationToken) :
+            return
+                cancellationToken.IsCancellationRequested ? ValueTask.FromCanceled(cancellationToken) :
+                buffer.IsEmpty ? default :
```
Comment (ViktorHofer) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:803]: This is a bit hard to read outside of an IDE. Consider adding parentheses around the second inline if else statement.

**Review by ViktorHofer — APPROVED** [review]

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:515]: Technically I think it's more correct as it currently is, even though it reads awkwardly.  We didn't pass compressed bytes: we passed uncompressed bytes, and we're then writing out the result, so we're writing, compressed, the bytes we already passed to the defaltor.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:803]: I read this as:
```C#
if (cancellationToken.IsCancellationRequested) return ValueTask.FromCanceled(cancellationToken);
if (buffer.IsEmpty) return default;
return Core(buffer, cancellationToken);
```
Where are you recommending parens go?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs (hunk 2 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs:515]: Oh! I see it now. it's weird but now it makes sense.

**Comment by ericstj** [comment]
@stephentoub did you create a breaking change doc for this?

**Pull request merged**